tools/apps/src/bin/quiche-client.rs

```diff
@@ -0,0 +1,790 @@
+                    break;
+                },
+
+                Err(e) => {
+                    error!("send failed: {:?}", e);
+
+                    conn.close(false, 0x1, b"fail").ok();
+                    break;
+                },
+            };
+
+            if let Err(e) = socket.send(&out[..write]) {
+                if e.kind() == std::io::ErrorKind::WouldBlock {
+                    debug!("send() would block");
+                    break;
+                }
+
+                panic!("send() failed: {:?}", e);
+            }
+
+            debug!("written {}", write);
+        }
+
+        if conn.is_closed() {
+            info!("connection closed, {:?}", conn.stats());
+
+            report_http_incomplete(&http_conn, &req_start);
+
+            break;
+        }
+    }
+}
+
+// Makes a buffered writer for the response body returned by the target URL. The
+// file will have the same name as the resource's last path segment value.
+// Multiple requests for the same URL are indiated by the value of `cardinal`,
+// any value "N" greater than 1, will cause ".N" to be appended to the filename.
+fn make_fd(
+    url: &url::Url, target_path: &str, cardinal: u64,
+) -> std::io::BufWriter<std::fs::File> {
+    let resource = url.path_segments().map(|c| c.collect::<Vec<_>>()).unwrap();
+
+    let mut path = format!("{}/{}", target_path, resource.iter().last().unwrap());
+
+    if cardinal > 1 {
+        path = format!("{}.{}", path, cardinal);
+    }
+
+    match std::fs::File::create(&path) {
+        Ok(f) => std::io::BufWriter::new(f),
+        Err(e) => panic!("Bad times: {}", e),
+    }
+}
+
+fn report_http_incomplete(
+    http_conn: &Option<HttpConn>, start: &std::time::Instant,
+) {
+    let (reqs_complete, reqs_count) = match &http_conn {
+        Some(HttpConn::Http09Conn {
+            reqs_complete,
+            reqs,
+            ..
+        }) => (*reqs_complete, reqs.len()),
+
+        Some(HttpConn::Http3Conn {
+            reqs_complete,
+            reqs,
+            ..
+        }) => (*reqs_complete, reqs.len()),
+
+        None => (0, 0),
+    };
+
+    if reqs_complete != reqs_count {
+        error!(
+            "connection timed out after {:?} and only completed {}/{} requests",
+            start.elapsed(),
+            reqs_complete,
+            reqs_count
+        );
+    }
+}
+
+// Represents an HTTP/0.9 formatted request.
+struct Http09Request {
+    url: url::Url,
+    cardinal: u64,
+    request_line: String,
+}
+
+// Represents an HTTP/3 formatted request.
+struct Http3Request {
+    url: url::Url,
+    cardinal: u64,
+    hdrs: Vec<quiche::h3::Header>,
+}
+
+// Used for application tracking of HTTP requests and responses.
+#[allow(clippy::large_enum_variant)]
+enum HttpConn {
+    Http09Conn {
+        stream_id: u64,
+        reqs_sent: usize,
+        reqs_complete: usize,
+        reqs: Vec<Http09Request>,
+        fds: HashMap<u64, std::io::BufWriter<std::fs::File>>,
```

Comment:
Why isn't the "fd" just part of the request struct? Also, would propbably call it something else, like, "writer" or something.

Reply:
I think this is down to being an afterthought feature mixed with wanting to keep a sane order for issuing requests. When not doing anything with the responses we didn't need any look up based on their stream ID. Let me take another look at this and see if I can find something more logical.